apps/src/netsim/NetSimVisualization.js

```diff
@@ -0,0 +1,475 @@
+ * @param {number} entityID - ID, with corresponds to NetSimEntity.entityID
+ * @returns {NetSimVizEntity} or null if not found
+ */
+NetSimVisualization.prototype.getEntityByID = function (entityType, entityID) {
+  return this.entities_.reduce(function (prev, cur) {
+    if (prev) {
+      return prev;
+    } else if (cur instanceof entityType && cur.id === entityID) {
+      return cur;
+    }
+    return null;
+  }, null);
+};
+
+/**
+ * Gets the set of VizWires directly attached to the given VizNode, (either
+ * on the local end or remote end)
+ * @param {NetSimVizNode} vizNode
+ * @returns {Array.<NetSimVizWire>} the attached wires
+ */
+NetSimVisualization.prototype.getWiresAttachedToNode = function (vizNode) {
+  return this.entities_.filter(function (entity) {
+    return entity instanceof NetSimVizWire &&
+        (entity.localVizNode === vizNode || entity.remoteVizNode === vizNode);
+  });
+};
+
+/**
+ * Handle notification that node table contents have changed.
+ * @param {Array.<Object>} rows - node table rows
+ * @private
+ */
+NetSimVisualization.prototype.onNodeTableChange_ = function (rows) {
+  // Convert rows to correctly-typed objects
+  var tableNodes = netsimUtils.nodesFromRows(this.shard_, rows);
+
+  // 1. Kill nodes from the visualization that are no longer in the table.
+  this.entities_.filter(function (vizEntity) {
+    return vizEntity instanceof NetSimVizNode;
+  }).filter(function (vizNode) {
+    return !tableNodes.some(function (node) {
+      return node.entityID === vizNode.id;
+    });
+  }).forEach(function (vizNode) {
+    vizNode.kill();
+  });
+
+  // 2. Add new nodes from the table into the visualization
+  tableNodes.forEach(function (node) {
+    var vizNode = this.getEntityByID(NetSimVizNode, node.entityID);
+    if (vizNode) {
+      vizNode.configureFrom(node);
+    } else {
+      vizNode = new NetSimVizNode(node);
+      vizNode.snapToPosition(
+          Math.random() * this.visualizationWidth - (this.visualizationWidth / 2),
+          Math.random() * this.visualizationHeight - (this.visualizationHeight / 2));
+      this.entities_.push(vizNode);
+      this.svgRoot_.find('#background_group').prepend(vizNode.getRoot());
+    }
+  }, this);
+};
+
+/**
+ * Handle notification taht wire table contents have changed.
+ * @param {Array.<Object>} rows - wire table rows
+ * @private
+ */
+NetSimVisualization.prototype.onWireTableChange_ = function (rows) {
+  // Convert rows to correctly-typed objects
+  var tableWires = rows.map(function (row) {
+    return new NetSimWire(this.shard_, row);
+  }.bind(this));
+
+  // 1. Kill wires that are no longer in the table
+  this.entities_.filter(function (vizEntity) {
+    return vizEntity instanceof NetSimVizWire;
+  }).filter(function (vizWire) {
+    return !tableWires.some(function (wire) {
+      return wire.entityID === vizWire.id;
+    });
+  }).forEach(function (vizWire) {
+    vizWire.kill();
+  });
```

Comment:
Another case where it seems like we could do this in one loop. Also, is this the same code? Does it deserve its own function?